### PR TITLE
fix(vanilla): should recompute state after onMount

### LINKED ISSUE #2098: atomWithStorage issue since 2.3.0+
This is a tracking issue for #2095. See discussions there.

Reproduction: https://codesandbox.io/s/sharp-lamport-k9s8mp?file=/src/App.tsx

I think it's a bug in https://github.com/pmndrs/jotai/pull/2029 and it's not specific to atomWithStorage.

Tasks:
1. Create a failing test (without atomWithStorage, hopefully only with store api)
2. Fix it

### PR DESCRIPTION
**Pull request by tmkx**
## Related Issues or Discussions

Fixes #2098 

## Check List

- [x] `yarn run prettier` for formatting code and docs

### CODE REVIEW

**Review by dai-shi — CHANGES_REQUESTED** [review]
Thanks for working on it!

Can you add a test with store api, without atomWithStorage, and without React?

File: tests/react/vanilla-utils/atomWithStorage.test.tsx [hunk 1]
```diff
@@ -445,6 +445,62 @@ describe('atomWithStorage (with browser storage)', () => {
...
+
+  it("should recompute dependents' state after onMount (#2098)", async () => {
+    const store = createStore()
+    const mockNativeStorage = Object.create(window.Storage.prototype)
+    let currentValue: string | null = 'true'
+    mockNativeStorage.setItem = vi.fn(
+      (_key: string, value: string) => (currentValue = value)
+    )
+    mockNativeStorage.getItem = vi.fn(() => currentValue)
+    mockNativeStorage.removeItem = vi.fn(() => (currentValue = null))
```
Comment (dai-shi) [inline tests/react/vanilla-utils/atomWithStorage.test.tsx:457]: Can it be something like this?
```suggestion
    let currentValue: string | null = 'true'
    const mockNativeStorage = {
      setItem: vi.fn(
      (_key: string, value: string) => (currentValue = value)
    ),
      getItem: vi.fn(() => currentValue),
      removeItem: vi.fn(() => (currentValue = null)),
    }
```

File: tests/react/vanilla-utils/atomWithStorage.test.tsx [hunk 2]
```diff
@@ -445,6 +445,62 @@ describe('atomWithStorage (with browser storage)', () => {
...
+
+    const isLoggedAtom = atom(false)
+    const _isDevModeState = atomWithStorage(
+      '_isDevModeState',
```
Comment (dai-shi) [inline tests/react/vanilla-utils/atomWithStorage.test.tsx:461]: We reserve `_foo` for unused names:
```suggestion
    const isDevModeState = atomWithStorage(
      'isDevModeState',
```

File: tests/react/vanilla-utils/atomWithStorage.test.tsx [hunk 3]
```diff
@@ -445,6 +445,62 @@ describe('atomWithStorage (with browser storage)', () => {
...
+
+    const { getByRole } = render(
+      <StrictMode>
+        <DummyComponent />
+      </StrictMode>
+    )
+
+    act(() => store.set(isLoggedAtom, true))
+
+    const $checkbox = getByRole('checkbox') as HTMLInputElement
```
Comment (dai-shi) [inline tests/react/vanilla-utils/atomWithStorage.test.tsx:495]: Just a preference:
```suggestion
    const checkbox = getByRole('checkbox') as HTMLInputElement
```

File: src/vanilla/store.ts [hunk 4]
```diff
@@ -594,6 +583,17 @@ export const createStore = () => {
     if (import.meta.env?.MODE !== 'production') {
       mountedAtoms.add(atom)
     }
+    // mount dependencies before mounting self
```
Comment (dai-shi) [inline src/vanilla/store.ts:586]: I wonder if we should mount dependencies before mounting self or after mounting self. There was a failing test in #2029 AFAIR, but hmm maybe it's fixed afterwards.

**Review by dai-shi — COMMENTED** [review]
The tricky thing is CI tests.

File: tests/react/vanilla-utils/atomWithStorage.test.tsx (hunk 1 above)
Comment (tmkx) [reply tests/react/vanilla-utils/atomWithStorage.test.tsx]: updated, it's copied from https://github.com/pmndrs/jotai/blob/2ec0bb96533f827b2b6f38456bbeab9cc542eddc/tests/react/vanilla-utils/atomWithStorage.test.tsx#L380-L383, should I update it too?

**Review by dai-shi — APPROVED** [review]
It looks good. Nice work.

Let me try and see if I can refactor.

**Comment by dai-shi** [comment]
I guess the fundamental issue is that dependent graph isn't ready when onMount is invoked. So, the queue is introduced. Not really sure if it's the ideal solution, but it works so far.

**Comment by fturmel** [comment]
Thank you both!

**Pull request merged**